client/src/main/java/io/split/client/JsonLocalhostSplitChangeFetcher.java

```diff
@@ -24,49 +24,61 @@ public class JsonLocalhostSplitChangeFetcher implements SplitChangeFetcher {
-            splitChange.ruleBasedSegments.d = new ArrayList<>();
-            splitChange.ruleBasedSegments.t = -1;
-            splitChange.ruleBasedSegments.s = -1;
-            return processSplitChange(splitChange, since);
+            return processSplitChange(splitChange, since, sinceRBS);
         } catch (Exception e) {
             throw new IllegalStateException("Problem fetching splitChanges: " + e.getMessage(), e);
         }
     }
 
-    private SplitChange processSplitChange(SplitChange splitChange, long changeNumber) throws NoSuchAlgorithmException {
+    private SplitChange processSplitChange(SplitChange splitChange, long changeNumber, long changeNumberRBS) throws NoSuchAlgorithmException {
         SplitChange splitChangeToProcess = LocalhostSanitizer.sanitization(splitChange);
         // if the till is less than storage CN and different from the default till ignore the change
-        if (splitChangeToProcess.featureFlags.t < changeNumber && splitChangeToProcess.featureFlags.t != -1) {
+        if (checkExitConditions(splitChangeToProcess.featureFlags,  changeNumber) ||
+            checkExitConditions(splitChangeToProcess.ruleBasedSegments,  changeNumberRBS)) {
             _log.warn("The till is lower than the change number or different to -1");
             return null;
         }
-        String splitJson = splitChange.featureFlags.d.toString();
-        MessageDigest digest = MessageDigest.getInstance("SHA-1");
-        digest.reset();
-        digest.update(splitJson.getBytes());
-        // calculate the json sha
-        byte [] currHash = digest.digest();
+        byte [] currHashFeatureFlags = getStringDigest(splitChange.featureFlags.d.toString());
+        byte [] currHashRuleBasedSegments = getStringDigest(splitChange.ruleBasedSegments.d.toString());
         //if sha exist and is equal to before sha, or if till is equal to default till returns the same segmentChange with till equals to storage CN
-        if (Arrays.equals(lastHash, currHash) || splitChangeToProcess.featureFlags.t == -1) {
+        if (Arrays.equals(lastHashFeatureFlags, currHashFeatureFlags) || splitChangeToProcess.featureFlags.t == -1) {
             splitChangeToProcess.featureFlags.t = changeNumber;
         }
-        lastHash = currHash;
+        if (Arrays.equals(lastHashRuleBasedSegments, currHashRuleBasedSegments) || splitChangeToProcess.ruleBasedSegments.t == -1) {
+            splitChangeToProcess.ruleBasedSegments.t = changeNumberRBS;
+        }
+
+        lastHashFeatureFlags = currHashFeatureFlags;
+        lastHashRuleBasedSegments = currHashRuleBasedSegments;
         splitChangeToProcess.featureFlags.s = changeNumber;
+        splitChangeToProcess.ruleBasedSegments.s = changeNumberRBS;
+
         return splitChangeToProcess;
     }
+
+    private <T> boolean checkExitConditions(ChangeDto<T> change, long cn) {
+        return change.t < cn && change.t != -1;
+    }
```

Comment:
we are repeating this validation in more than one place. I suggest to moving it to a utility

Reply:
will fix